book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -167,6 +167,25 @@ chains, so that the map ordering is the ordering of best to worst chains.
 
 - XXX fill in details on exact types
 
+- **Chain**: `(im::OrdMap<block::Height, Arc<Block>>, HashSet<Nullifier>, HashSet<block::Hash>)`
```

Comment:
Do we need to keep a nullifier set for each block in the chain, to support forks at arbitrary points in the chain?

That is:
```suggestion
- **Chain**: `(im::OrdMap<block::Height, (Arc<Block>, HashSet<Nullifier>)>, HashSet<block::Hash>)`
```

If not, how to we calculate the nullifier set during a fork?

This is related to my other question - I think our code will be simpler and more reliable if we make push and fork the same operation, even if there's a slightly higher storage and CPU cost.

Reply:
🤔 good point

Reply:
does https://github.com/ZcashFoundation/zebra/pull/902#discussion_r485784361 help?